lazy_theta_star_b/src/lazy_theta_star_b.cpp

```diff
@@ -0,0 +1,529 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Anshumaan Singh
+ *********************************************************************/
+#include "lazy_theta_star_b/lazy_theta_star_b.h"
+#include "nav2_util/node_utils.hpp"
+#include "chrono"
+
+namespace lazy_theta_star_b {
+    void LazyThetaStarB::configure(rclcpp_lifecycle::LifecycleNode::SharedPtr parent,
+                                   std::string name, std::shared_ptr<tf2_ros::Buffer> tf,
+                                   std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros) {
+        node_ = parent;
+        name_ = name;
+        tf_ = tf;
+        costmap_ = costmap_ros->getCostmap();
+        global_frame_ = costmap_ros->getGlobalFrameID();
+
+        nav2_util::declare_parameter_if_not_declared(
+                node_, name_ + ".how_many_corners", rclcpp::ParameterValue(
+                        8));
+
+        node_->get_parameter(name_ + ".how_many_corners", how_many_corners_);
+
+        if (how_many_corners_ != 8)
+            how_many_corners_ = 4;
+
+        nav2_util::declare_parameter_if_not_declared(
+                node_, name_ + ".interpolation_dist", rclcpp::ParameterValue(
+                        0.1));
+
+        node_->get_parameter(name_ + ".interpolation_dist", interpolation_dist);
+
+    }
+
+    void LazyThetaStarB::cleanup() {
+        RCLCPP_INFO(
+                node_->get_logger(), "CleaningUp plugin %s of type LazyThetaStarB",
+                name_.c_str());
+    }
+
+    void LazyThetaStarB::activate() {
+        RCLCPP_INFO(
+                node_->get_logger(), "Activating plugin %s of type LazyThetaStarB",
+                name_.c_str());
+    }
+
+    void LazyThetaStarB::deactivate() {
+        RCLCPP_INFO(
+                node_->get_logger(), "Deactivating plugin %s of type LazyThetaStarB",
+                name_.c_str());
+    }
+
+    nav_msgs::msg::Path LazyThetaStarB::createPlan(
+            const geometry_msgs::msg::PoseStamped &start,
+            const geometry_msgs::msg::PoseStamped &goal) {
+
+
+           RCLCPP_INFO(
+                node_->get_logger(), "GOT THE SOURCE AND DESTINATION ------ (%f, %f) && (%f, %f)",
+               start.pose.position.x, start.pose.position.y, goal.pose.position.x, goal.pose.position.y);
+
+        costmap_->worldToMap(start.pose.position.x, start.pose.position.y, reinterpret_cast<unsigned int &>(src.x),
+                             reinterpret_cast<unsigned int &>(src.y));
+
+
+        costmap_->worldToMap(goal.pose.position.x, goal.pose.position.y, reinterpret_cast<unsigned int &>(dst.x),
+                             reinterpret_cast<unsigned int &>(dst.y));
+        src.x = int(src.x);
+        src.y = int(src.y);
+
+        dst.x = int(dst.x);
+        dst.y = int(dst.y);
+
+        data.clear();
+        pq.clear();
+        path.clear();
+        posn.clear();
+        lethal_cost = LETHAL_COST;
```

Comment:
Lethal cost set to 100? for the costmap, the lethal cost is 253, 100 is just an inflated cost that's valid

Reply:
If set to 253, it gives a path that might require the robot to go near the edges of higher lethal_cost, so for testing I had kept it at 100, to be able to continuously test it

Reply:
Sure, your other option is to make the heuristic have some weight on cost. What's your heuristic based on? Just distance? You could add some additional cost on the edge dependent on value at the cell too. That's what NavFn / Hybrid-A* / other planners typically do. Then there's a weight parameter to set the scale of the cost applied by that heuristic. Higher means stay away from cost and lower means near is fine as long as non-lethal.

Reply:
I am not sure whether such a heuristic will work well for this particular algorithm, which has to find any-angle paths, in my opinion still having inflated costs on the costmap around the obstacles, which won't be expanded at all by the algorithm seems better, as the user can set the inflation radius as they want (albeit it would require more inflation than usual).